### PR TITLE
Create test page

### PR DESCRIPTION
**Pull request by Junk-debug**

### CODE REVIEW

**Comment by Junk-debug** [comment]
https://junk-debug.github.io/react_Examix/

File: client/src/components/TestInfo.tsx [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+import ImageLinkUploader from '../dev/components/ImageLinkUploader';
+import { useCreateTest } from '../pages/CreateTestPage/CreateTestContext';
+
+interface Props extends BoxProps {}
+
+const TestInfo: React.FC<Props> = ({ sx, ...rest }) => {
+  const {
+    register,
+    formState: { errors },
+  } = useFormContext<CreateTestForm>();
```
Comment (Softjey) [inline client/src/components/TestInfo.tsx:14]: You use Form context you have to rename this component. Keep all similar components somewhere together in one folder.

File: client/src/components/TestInfo.tsx [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+  const { loading } = useCreateTest();
+
+  return (
+    <Box {...rest} display="flex" flexDirection="column" gap="24px" sx={{ width: '100%', ...sx }}>
+      <ImageLinkUploader
+        sx={{
+          alignSelf: 'start',
```
Comment (Softjey) [inline client/src/components/TestInfo.tsx:22]: use flex-start instead

File: client/src/components/TestInfo.tsx (hunk 1 above)
Comment (Softjey) [inline client/src/components/TestInfo.tsx:14]: create custom typed hook, useCreateTestForm

File: client/src/components/UI/ErrorSnackBar.tsx [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    if (reason === 'clickaway') {
+      return;
+    }
+
+    onClose();
+  };
+
+  return (
+    <Snackbar
+      sx={{ width: 400 }}
```
Comment (Softjey) [inline client/src/components/UI/ErrorSnackBar.tsx:21]: also spread sx prop

File: client/src/components/UI/QuestionTypeSelect.tsx [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+const QuestionTypeSelect: React.FC<Props> = ({ ...rest }) => {
+  return (
+    <TextField
+      defaultValue={QuestionType.SINGLE_CHOICE}
+      {...rest}
+      size="small"
+      select
+      sx={{ width: '158px' }}
+    >
+      {(Object.keys(QuestionType) as Array<keyof typeof QuestionType>).map((type) => (
```
Comment (Softjey) [inline client/src/components/UI/QuestionTypeSelect.tsx:16]: don't use Object keys, use Object values instead

File: client/src/components/UI/QuestionsAutoComplete/renderOption.tsx [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  { ...rest },
+  question: Question,
+) => {
+  return (
+    <LinkListOption
+      key={question.id}
+      title={trim(question.title, 60)}
+      subtitle={
+        <Stack direction="row" alignItems="center" spacing={1}>
+          {/* <Chip color="primary" size="small" label={question.id} /> */}
```
Comment (Softjey) [inline client/src/components/UI/QuestionsAutoComplete/renderOption.tsx:20]: delete this comment

File: client/src/components/UI/QuestionsAutoComplete/renderOption.tsx [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+              variant="chip"
+              chipColor="default"
+              chipVariant="outlined"
+              subject={question.subject}
+            />
+          )}
+
+          <Chip size="small" label={underscoreToUpperToSentence(question.type)} />
+
+          {/* <Chip
```
Comment (Softjey) [inline client/src/components/UI/QuestionsAutoComplete/renderOption.tsx:32]: this also delete

File: client/src/components/items/FormAnswerItem.tsx [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+  onDelete,
+  onCheckBoxClick,
+  isFromServer,
+}) => {
+  const {
+    register,
+    watch,
+    setValue,
+    formState: { errors },
+  } = useFormContext<CreateTestForm>();
```
Comment (Softjey) [inline client/src/components/items/FormAnswerItem.tsx:31]: create custom hook and use it everywhere

File: client/src/components/items/FormAnswerItem.tsx [hunk 8]
```diff
@@ -0,0 +1,93 @@
+import { useFormContext } from 'react-hook-form';
+import { Box, TextField } from '@mui/material';
+import { CreateTestForm } from '../../schemas/createTestFormValidationSchemas';
+import QuestionType from '../../types/api/enums/Type';
+import CloseButton from '../UI/buttons/CloseButton';
+import RadioCheckBox from '../UI/RadioCheckBox';
+import { useCreateTest } from '../../pages/CreateTestPage/CreateTestContext';
+
+interface Props {
```
Comment (Softjey) [inline client/src/components/items/FormAnswerItem.tsx:9]: use Props extends BoxProps

File: client/src/components/items/FormAnswerItem.tsx [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+  return (
+    <Box sx={{ display: 'flex', alignItems: 'start' }}>
+      <RadioCheckBox
+        {...register(`questions.${questionIndex}.answers.${answerIndex}.isCorrect`)}
+        type={type}
+        ref={null}
+        /* this is neccessary because mui checkbox component
+        is not working correctly with react-hook-form */
+        checked={watch(`questions.${questionIndex}.answers.${answerIndex}.isCorrect`)}
```
Comment (Softjey) [inline client/src/components/items/FormAnswerItem.tsx:43]: don't hardcode watch in props, make a variable, and change this in all files

File: client/src/components/items/FormAnswerItem.tsx [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+        onChange={(e) => {
```
Comment (Softjey) [inline client/src/components/items/FormAnswerItem.tsx:44]: if function bigger than 1 line, don't hardcode it, change this in all files

File: client/src/components/items/FormAnswerItem.tsx [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+        error={!!errors.questions?.[questionIndex]?.answers?.[answerIndex]?.title}
+        helperText={errors.questions?.[questionIndex]?.answers?.[
+          answerIndex
+        ]?.title?.message?.toString()}
+        autoComplete="off"
+        size="small"
+        sx={{
+          minWidth: '200px',
+          '&:hover .icon-button':
+            loading || isFromServer
```
Comment (Softjey) [inline client/src/components/items/FormAnswerItem.tsx:62]: make variable for loading || isFromServer and change it everywhere

File: client/src/dev/components/AnswersGroup.tsx [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+
+  const { loading } = useCreateTest();
+
+  const isShowAddButton = fields.length >= 6 || isFromServer;
+
+  const updateCorrect = (index: number) => {
+    const answers = watch(`questions.${questionIndex}.answers`);
+
+    if (questionType === QuestionType.SINGLE_CHOICE) {
+      if (answers.some((item, i) => item.isCorrect && i !== index)) {
```
Comment (Softjey) [inline client/src/dev/components/AnswersGroup.tsx:43]: make var for that, don't hardcode in if statement

File: client/src/dev/components/AnswersGroup.tsx [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+        // If any checkbox is checked and it's not the current one, uncheck all checkboxes
+        setValue(
+          `questions.${questionIndex}.answers`,
+          answers.map((item, i) => {
```
Comment (Softjey) [inline client/src/dev/components/AnswersGroup.tsx:47]: and for that also

Comment (Junk-debug) [reply client/src/dev/components/AnswersGroup.tsx:47]: for what?
setValue path?

File: client/src/services/Router/router.tsx [hunk 14]
```diff
@@ -80,7 +82,9 @@ const router = createRouter([
     path: Routes.CREATE_TEST,
     element: (
       <Authenticated>
-        <HomeLayout>Create test page</HomeLayout>
+        <CreateTestProvider>
```
Comment (Softjey) [inline client/src/services/Router/router.tsx:85]: Don't use provider here, it is a part of CreateTestPage

**Review by Softjey — COMMENTED** [review]
Add move all componets and another files from dev folder!

File: client/src/dev/components/TimeLimitPicker.tsx [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+        const { onBlur, onChange, ref, value, disabled } = field;
+        return (
+          <TimePicker
+            sx={{
+              maxWidth: '105px',
+              '.MuiInputBase-input': {
+                paddingTop: 1,
+                paddingBottom: 1,
+                '& ~ fieldset': {
+                  borderColor: error ? 'red !important' : 'auto',
```
[the remark this answers is not in the document]
Comment (Junk-debug) [reply client/src/dev/components/TimeLimitPicker.tsx]: how to use it?

File: client/src/pages/CreateTestPage/CreateTestPage.tsx [hunk 16]
```diff
@@ -0,0 +1,227 @@
...
+
+    const filteredQuestions = getFilteredQuestions(data);
+
+    const questionsFromServer = data.questions.filter((question) => question.isFromServer);
+
+    console.log('data to send', filteredQuestions);
+
+    createQuestions(filteredQuestions, {
+      onError: (err) => {
+        console.dir(err);
```
[the remark this answers is not in the document]
Comment (Junk-debug) [reply client/src/pages/CreateTestPage/CreateTestPage.tsx:96]: this is uneccessary because we have error handling via snackbar

**Review by Softjey — COMMENTED** [review]
move hook useIds into hooks folder, move schemas into CreateTestForm folder

**Pull request merged**